### PR TITLE
Ux updates

### PR DESCRIPTION
**Pull request by winniequinn**
I believe @gioneill is still working on this, but I am creating a PR just so I can review.

https://github.com/NYPL-Simplified/Simplified-iOS/issues/420

fixes #1

### CODE REVIEW

File: NYPLCardCreator/Classes/AddressViewController.swift [hunk 1]
```diff
@@ -3,16 +3,6 @@ import UIKit
 /// This class is used for allowing the user to enter an address.
 final class AddressViewController: FormTableViewController {
   
-  private static let regions: [String] = {
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/AddressViewController.swift]: If we're definitely removing the picker view, be sure to also remove "regions.json" from the project so it's not hanging around doing nothing.

File: NYPLCardCreator/Classes/AlternativeAddressesViewController.swift [hunk 2]
```diff
@@ -32,16 +32,31 @@ final class AlternativeAddressesViewController: TableViewController {
   
   override func viewDidLoad() {
     super.viewDidLoad()
+    
+    let headerView = UIView(frame: CGRect(x: 0, y: 0, width: self.view!.bounds.size.width, height: 80.0))
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/AlternativeAddressesViewController.swift]: All of the sizes of views in the card creator were previously set without any fixed numbers like `80.0`. This is important for supporting things like dynamic text sizing and text of different lengths depending on the current locale. I would rework this so you do not need to rely on a magic value for the size of the `UIView`. If you need some advice on how to do this, please poke me!

File: NYPLCardCreator/Classes/AlternativeAddressesViewController.swift [hunk 3]
```diff
@@ -32,16 +32,31 @@ final class AlternativeAddressesViewController: TableViewController {
...
+    let headerLabel = UILabel()
+    headerView.addSubview(headerLabel)
+    
+    headerLabel.autoPinEdgesToSuperviewEdgesWithInsets(UIEdgeInsetsMake(2.0, 22.0, 2.0, 22.0))
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/AlternativeAddressesViewController.swift]: The same advice regarding magic numbers goes here too. Apple provides a concept of view margins that you can take advantage of via PureLayout that should let you do this in a cleaner way.

File: NYPLCardCreator/Classes/AlternativeAddressesViewController.swift [hunk 4]
```diff
@@ -32,16 +32,31 @@ final class AlternativeAddressesViewController: TableViewController {
...
+    headerLabel.numberOfLines = 2
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/AlternativeAddressesViewController.swift]: Again, fixing this at `2` makes things fragile given that the length of the text will change with different localizations.

File: NYPLCardCreator/Classes/AlternativeAddressesViewController.swift [hunk 5]
```diff
@@ -32,16 +32,31 @@ final class AlternativeAddressesViewController: TableViewController {
...
+    headerLabel.textColor = UIColor.darkGrayColor()
+    headerLabel.textAlignment = .Center
+    
+    self.tableView.tableHeaderView = headerView
+    
     switch self.addressStep {
     case .Home:
+      headerLabel.text = "Select the correct home address, or go back to make changes."
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/AlternativeAddressesViewController.swift]: You need to be using `NSLocalizedString` here as you would with any other user-visible string.

File: NYPLCardCreator/Classes/ConfirmValidAddressViewController.swift [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    
+    // Estimated cell height obtained via debugging. This must be set in order for the cells
+    // to be sized automatically via `UITableViewAutomaticDimension` (which is the default).
+    self.tableView.estimatedRowHeight = 104
+  }
+  
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    
+    let headerView = UIView(frame: CGRect(x: 0, y: 0, width: self.view!.bounds.size.width, height: 80.0))
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/ConfirmValidAddressViewController.swift]: Magic numbers, et cetera!

File: NYPLCardCreator/Classes/ConfirmValidAddressViewController.swift [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    let headerLabel = UILabel()
+    headerView.addSubview(headerLabel)
+    
+    headerLabel.autoPinEdgesToSuperviewEdgesWithInsets(UIEdgeInsetsMake(2, 22, 2, 22))
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/ConfirmValidAddressViewController.swift]: Magic numbers, et cetera!

File: NYPLCardCreator/Classes/LocationViewController.swift [hunk 8]
```diff
@@ -33,6 +33,8 @@ final class LocationViewController: UIViewController {
     self.title = NSLocalizedString(
       "Location Check",
       comment: "A title telling the user the app needs to check their location")
+    let backButton = UIBarButtonItem(title: "Location", style: .Bordered, target: nil, action: nil)
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/LocationViewController.swift]: You need to use `NSLocalizedString` for "Location" here.

Comment (winniequinn) [inline NYPLCardCreator/Classes/LocationViewController.swift]: Why is a target and action not needed for the bar button?

File: NYPLCardCreator/Classes/NameAndEmailViewController.swift [hunk 9]
```diff
@@ -77,10 +77,23 @@ final class NameAndEmailViewController: FormTableViewController {
       animated: true)
   }
   
+  private func emailIsValid() -> Bool {
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/NameAndEmailViewController.swift]: As we discussed, email addresses can be really complicated and this regex will exclude valid email addresses. See https://davidcel.is/posts/stop-validating-email-addresses-with-regex/ for some thoughts on this topic. Personally, I would just go back to the old approach of making sure `@` is present and that's it. The user will get a chance to review their email address at the end anyway.

File: NYPLCardCreator/Classes/UserSummaryViewController.swift [hunk 10]
```diff
@@ -0,0 +1,239 @@
+import UIKit
+
+/// This class is used for summarizing the user's details before
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/UserSummaryViewController.swift]: This is extremely anal, but this comment should end with a period to be consistent with Apple's conventions for class descriptions.

File: NYPLCardCreator/Classes/UserSummaryViewController.swift [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+      "Username",
+      "Pin"
+    ]
+    if (self.schoolOrWorkAddress != nil) {
+      self.cells.insert(self.altAddressCell, atIndex: 1)
+      self.sectionHeaderTitles.insert("School or Work Address", atIndex: 1)
+    }
+    
+    super.init(style: .Grouped)
+    self.tableView.estimatedRowHeight = 104
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/UserSummaryViewController.swift]: This estimated row height is not correct for these cells. `104` was only used for cells that show addresses.

File: NYPLCardCreator/Classes/UserSummaryViewController.swift [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+    self.sectionHeaderTitles = [
+      "Home Address",
+      "Full Name",
+      "Email",
+      "Username",
+      "Pin"
+    ]
+    if (self.schoolOrWorkAddress != nil) {
+      self.cells.insert(self.altAddressCell, atIndex: 1)
+      self.sectionHeaderTitles.insert("School or Work Address", atIndex: 1)
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/UserSummaryViewController.swift]: This string needs to be localized.

File: NYPLCardCreator/Classes/UserSummaryViewController.swift [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+    self.prepareTableViewCells()
+  }
+  
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    self.title = NSLocalizedString(
+      "Review",
+      comment: "A title for a screen letting the user know they can review the information they have entered")
+    
+    let headerView = UIView(frame: CGRect(x: 0, y: 0, width: self.view!.bounds.size.width, height: 80.0))
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/UserSummaryViewController.swift]: Magic numbers.

File: NYPLCardCreator/Classes/UserSummaryViewController.swift [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+    let headerLabel = UILabel()
+    headerView.addSubview(headerLabel)
+    
+    headerLabel.autoPinEdgesToSuperviewEdgesWithInsets(UIEdgeInsetsMake(2, 22, 2, 22))
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/UserSummaryViewController.swift]: Magic numbers.

File: NYPLCardCreator/Classes/UserSummaryViewController.swift [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+    self.tableView.userInteractionEnabled = false
+  }
+  
+  private func prepareTableViewCells() {
+    for cell in self.cells {
+      cell.backgroundColor = UIColor.clearColor()
+      self.tableView.separatorStyle = .None
+      
+      if let addressViewCell = cell as? AddressCell {
+        //style as an address cell
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/UserSummaryViewController.swift]: I assume additional work is pending here.

File: NYPLCardCreator/Classes/UserSummaryViewController.swift [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+  func numberOfSectionsInTableView(tableView: UITableView) -> Int {
+    return self.cells.count
+  }
+  
+  override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+    return self.cells[indexPath.section]
+  }
+  
+  func tableView(tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
+    return 1.0
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/UserSummaryViewController.swift]: Surely this cannot be correct! Is there a reason we return a height for footers here?

File: NYPLCardCreator/Classes/UserSummaryViewController.swift [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+  }
+  
+  func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+    return self.sectionHeaderTitles[section]
+  }
+  
+  // MARK: -
+  
+  @objc private func createPatron() {
+//    self.view.endEditing(false)
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/UserSummaryViewController.swift]: If we don't need this, just kill it.

File: NYPLCardCreator/Classes/ValidateAddressResponse.swift [hunk 18]
```diff
@@ -33,9 +33,9 @@ final class ValidateAddressResponse {
         guard
           let JSONObject = object as? [String: AnyObject],
           let addressJSON = JSONObject["address"] as? [String: AnyObject],
-          let address = Address.addressWithJSONObject(addressJSON),
-          let cardTypeString = JSONObject["card_type"] as? String
+          let address = Address.addressWithJSONObject(addressJSON)
           else { return nil }
+        let cardTypeString = JSONObject["card_type"] as? String
```
Comment (winniequinn) [inline NYPLCardCreator/Classes/ValidateAddressResponse.swift]: Why did we make this change here? I think we're supposed to always expect a response with a `"card_type"` of type `String`, so moving that out of the guard introduces the chance of a bad response from the server not being caught (as cardTypeString can now be `nil`).

File: NYPLCardCreator/Classes/AddressViewController.swift [hunk 19]
```diff
@@ -115,8 +95,9 @@ final class AddressViewController: FormTableViewController {
     self.cityCell.textField.keyboardType = .Alphabet
     self.cityCell.textField.autocapitalizationType = .Words
     
-    self.regionCell.textField.inputView = self.regionPickerView
-    self.regionCell.textField.inputAccessoryView = self.returnToolbar()
+    self.regionCell.textField.keyboardType = .Alphabet
+    self.regionCell.textField.autocorrectionType = .No
```
Comment (gioneill) [inline NYPLCardCreator/Classes/AddressViewController.swift]: I was thinking it'd be more of a annoyance for users trying to type two letter abbreviations, but I've been out-voted on this one!

File: NYPLCardCreator/Classes/NameAndEmailViewController.swift (hunk 9 above)
Comment (gioneill) [inline NYPLCardCreator/Classes/NameAndEmailViewController.swift]: By request, this has been reduced in strictness to only check for an '@' symbol and at least one 'dot' following it.

File: NYPLCardCreator/Classes/UserSummaryViewController.swift (hunk 15 above)
Comment (gioneill) [inline NYPLCardCreator/Classes/UserSummaryViewController.swift]: Not at this time. Thanks.

File: NYPLCardCreator/Classes/ValidateAddressResponse.swift (hunk 18 above)
Comment (gioneill) [inline NYPLCardCreator/Classes/ValidateAddressResponse.swift]: Found an example of "card_type" returning nil when out-of-state address was used.

File: NYPLCardCreator/Classes/UserSummaryViewController.swift (hunk 11 above)
Comment (gioneill) [inline NYPLCardCreator/Classes/UserSummaryViewController.swift]: The summary tableview has a mix of address cells and standard single-line cells. I wasn't able to get correct rendering with a value less than the estimated height for the address cells. If there's a better approach to this let me know.

**Pull request merged**